embassy-stm32/src/rcc/h.rs

```diff
@@ -194,14 +253,101 @@ pub(crate) unsafe fn init(config: Config) {
+                    w.set_ldoen(true);
+                    w.set_bypass(false);
+                });
+            }
+            SupplyConfig::LDO => {
+                PWR.cr3().modify(|w| {
+                    w.set_sden(false);
+                    w.set_ldoen(true);
+                    w.set_bypass(false);
+                });
+            }
+            SupplyConfig::DirectSMPS => {
+                PWR.cr3().modify(|w| {
+                    w.set_sdexthp(false);
+                    w.set_sden(true);
+                    w.set_ldoen(false);
+                    w.set_bypass(false);
+                });
+            }
+            SupplyConfig::SMPSLDO => {
+                PWR.cr3().modify(|w| {
+                    match config.smps_supply_voltage {
+                        Some(SMPSSupplyVoltage::V1_8) => {
+                            PWR.cr3().modify(|w| w.set_sdlevel(0b01));
+                        }
+                        Some(SMPSSupplyVoltage::V2_5) => {
+                            PWR.cr3().modify(|w| w.set_sdlevel(0b10));
+                        }
+                        None => {
+                            panic!("Supply configuration SMPSLDO requires a supply voltage to be set.");
+                        }
+                    }
+                    w.set_sdexthp(false);
+                    w.set_sden(true);
+                    w.set_ldoen(true);
+                    w.set_bypass(false);
+                });
+            }
+            SupplyConfig::SMPSExternalLDO => {
+                PWR.cr3().modify(|w| {
+                    match config.smps_supply_voltage {
+                        Some(SMPSSupplyVoltage::V1_8) => {
+                            PWR.cr3().modify(|w| w.set_sdlevel(0b01));
+                        }
+                        Some(SMPSSupplyVoltage::V2_5) => {
+                            PWR.cr3().modify(|w| w.set_sdlevel(0b10));
+                        }
+                        None => {
+                            panic!("Supply configuration SMPSExternalLDO requires a supply voltage to be set.");
+                        }
+                    }
+                    w.set_sdexthp(true);
+                    w.set_sden(true);
+                    w.set_ldoen(true);
+                    w.set_bypass(false);
+                });
+            }
+            SupplyConfig::SMPSExternalLDOBypass => {
+                PWR.cr3().modify(|w| {
+                    match config.smps_supply_voltage {
+                        Some(SMPSSupplyVoltage::V1_8) => {
+                            PWR.cr3().modify(|w| w.set_sdlevel(0b01));
+                        }
+                        Some(SMPSSupplyVoltage::V2_5) => {
+                            PWR.cr3().modify(|w| w.set_sdlevel(0b10));
+                        }
+                        None => {
+                            panic!("Supply configuration SMPSExternalLDOBypass requires a supply voltage to be set.");
```

Comment:
if you make the voltage a field in the `SupplyConfig` variants that need it, it becomes impossible to not supply it:

```rust
pub enum SupplyConfig {
    Default,
    LDO,
    DirectSMPS,
    SMPSLDO(SMPSSupplyVoltage),
    SMPSExternalLDO(SMPSSupplyVoltage),
    SMPSExternalLDOBypass(SMPSSupplyVoltage),
}
```